src/hooks/useCombobox/index.js

```diff
@@ -170,309 +174,339 @@ function useCombobox(userProps = {}) {
-      if (event.which === 229) {
-        return
-      }
-
-      if (isOpen && highlightedIndex > -1) {
+  const inputKeyDownHandlers = useMemo(
+    () => ({
+      ArrowDown(event) {
         event.preventDefault()
         dispatch({
-          type: stateChangeTypes.InputKeyDownEnter,
+          type: stateChangeTypes.InputKeyDownArrowDown,
+          shiftKey: event.shiftKey,
           getItemNodeFromIndex,
         })
-      }
-    },
-  }
-
-  // Event handlers.
-  const inputHandleKeyDown = event => {
-    const key = normalizeArrowKey(event)
-    if (key && inputKeyDownHandlers[key]) {
-      inputKeyDownHandlers[key](event)
-    }
-  }
-  const inputHandleChange = event => {
-    dispatch({
-      type: stateChangeTypes.InputChange,
-      inputValue: isReactNative
-        ? /* istanbul ignore next (react-native) */ event.nativeEvent.text
-        : event.target.value,
-    })
-  }
-  const inputHandleBlur = () => {
-    /* istanbul ignore else */
-    if (!mouseAndTouchTrackersRef.current.isMouseDown) {
-      dispatch({
-        type: stateChangeTypes.InputBlur,
-      })
-    }
-  }
-  const menuHandleMouseLeave = () => {
-    dispatch({
-      type: stateChangeTypes.MenuMouseLeave,
-    })
-  }
-  const itemHandleMouseMove = index => {
-    if (index === highlightedIndex) {
-      return
-    }
-    shouldScroll.current = false
-    dispatch({
-      type: stateChangeTypes.ItemMouseMove,
-      index,
-    })
-  }
-  const itemHandleClick = index => {
-    dispatch({
-      type: stateChangeTypes.ItemClick,
-      index,
-    })
-  }
-  const toggleButtonHandleClick = () => {
-    dispatch({
-      type: stateChangeTypes.ToggleButtonClick,
-    })
+      },
+      ArrowUp(event) {
+        event.preventDefault()
+        dispatch({
+          type: stateChangeTypes.InputKeyDownArrowUp,
+          shiftKey: event.shiftKey,
+          getItemNodeFromIndex,
+        })
+      },
+      Home(event) {
+        event.preventDefault()
+        dispatch({
+          type: stateChangeTypes.InputKeyDownHome,
+          getItemNodeFromIndex,
+        })
+      },
+      End(event) {
+        event.preventDefault()
+        dispatch({
+          type: stateChangeTypes.InputKeyDownEnd,
+          getItemNodeFromIndex,
+        })
+      },
+      Escape() {
+        dispatch({
+          type: stateChangeTypes.InputKeyDownEscape,
+        })
+      },
+      Enter(event) {
+        // if IME composing, wait for next Enter keydown event.
+        if (event.which === 229) {
+          return
+        }
+        const latestState = latest.current.state
 
-    if (!isOpen && inputRef.current) {
-      inputRef.current.focus()
-    }
-  }
+        if (latestState.isOpen && latestState.highlightedIndex > -1) {
+          event.preventDefault()
+          dispatch({
+            type: stateChangeTypes.InputKeyDownEnter,
+            getItemNodeFromIndex,
+          })
+        }
+      },
+    }),
+    [dispatch, latest],
+  )
```

Comment:
Git is funny... This is the end of the `inputKeyDownHandlers` useMemo memoization.